File: cpp/questions/canFinish.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        std::vector<std::vector<int>> graph(numCourses);
        for (auto& v: prerequisites) {
            graph[v[0]].push_back(v[1]);
        }
        std::vector<int> flag(numCourses);
        std::function<bool(int)> dfs = [&](int c) -> bool {
            if (flag[c] == 1) {
                return false;
            } else if (flag[c] == 2) {
                return true;
            }
            flag[c] = 1;
            for (auto rc: graph[c]) {
                if (!dfs(rc)) {
                    return false;
                }
            }
            flag[c] = 2;
            return true;
        };
        for (int i=0; i<numCourses; ++i) {
            if (!dfs(i)) {
                return false;
            }
        }
        return true;
    }
};

File: python/questions/canFinish.py
from typing import List

class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = [[] for _ in range(numCourses)]
        for cs in prerequisites:
            graph[cs[0]].append(cs[1])
        flag = [0 for _ in range(numCourses)]
        def dfs(c: int) -> bool:
            if flag[c] == 1:
                return False
            elif flag[c] == 2:
                return True
            flag[c] = 1
            for nc in graph[c]:
                if not dfs(nc):
                    return False
            flag[c] = 2
            return True
        for i in range(numCourses):
            if not dfs(i):
                return False
        return True
